### PR TITLE
Fix deadlock in RuntimeManager

### LINKED ISSUE #2691: Deadlock can occour if VarsWatcher (or other components) overload CheckinExpected components
Probably related: https://github.com/elastic/elastic-agent/issues/2628

The `runtimeComm` component has a `checkinExpected` channel that's used to track updates to the state of a unit:
https://github.com/elastic/elastic-agent/blob/a05e24ad067917c1525656f923cfe0d40b37ff36/pkg/component/runtime/runtime_comm.go#L92

If the buffer on this channel fills up and channel sends begin to block (which can happen in cases where unit config updates are happening faster than unit Checkin events, as is sometimes the case with k8s installs with container providers), this will cause the `manager.update()` method to block while it holds a mutex:

https://github.com/elastic/elastic-agent/blob/a05e24ad067917c1525656f923cfe0d40b37ff36/pkg/component/runtime/manager.go#L655-L657

As the `manager.update()` method tries to send to a component that also becomes blocked when the `checkinExpected` channel blocks here:
https://github.com/elastic/elastic-agent/blob/a05e24ad067917c1525656f923cfe0d40b37ff36/pkg/component/runtime/manager.go#L674

That `Update()` call is implemented here:
https://github.com/elastic/elastic-agent/blob/a05e24ad067917c1525656f923cfe0d40b37ff36/pkg/component/runtime/command.go#L241-L244

And the method that reads from that `compCh` channel, `commandRuntime.Run`, is blocked because of a call to CheckinExpected():

https://github.com/elastic/elastic-agent/blob/a05e24ad067917c1525656f923cfe0d40b37ff36/pkg/component/runtime/command.go#L149

Which is blocked by a send to the full `c.checkinExpected` channel:

https://github.com/elastic/elastic-agent/blob/a05e24ad067917c1525656f923cfe0d40b37ff36/pkg/component/runtime/runtime_comm.go#L171-L172C31

This becomes a problem when the `manager.getCertificate()` tries to acquire that mutex, which it never will:
https://github.com/elastic/elastic-agent/blob/a05e24ad067917c1525656f923cfe0d40b37ff36/pkg/component/runtime/manager.go#L807-L820

The `getCertificate()` callback is called during the gRPC TLS handshake. However, the gRPC `CheckinV2` call is the only codepath that drains the aforementioned `checkinExpected` channel, meaning that the agent is effectively deadlocked at this point.

There's a few ways to fix this:
1. change the mTLS system so we don't need a `getCertificate()` callback.
2. make the `runtime.Update()` method non-blocking.
3. make the `runtimeComm.CheckinExpected()` method non-blocking
4. Use a more complex queue setup to handle checkinExpected events from other components instead of just a buffered channel

## Reproducing

Reproducing this is a bit difficult, as it requires the agent component managers to send a large number of unit updates. In the wild, this is often caused by the VarsWatcher component, which gets updates from the Docker/k8s provider:
https://github.com/elastic/elastic-agent/blob/a05e24ad067917c1525656f923cfe0d40b37ff36/internal/pkg/agent/application/coordinator/coordinator.go#L692-L698

That call to `processVars` will eventually call `manager.Update()`, which in turn calls `runtime.Update()`, which will end up populating that channel via `comm.CheckinExpected()`. The best way to reproduce this is to run the agent in an active k8s environment with the various container providers enabled.

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Fixes a deadlock issue that can occur in the communication of updated components with the `RuntimeManager` and each of the `CommandRuntime` and `ServiceRuntime` component managers.

## Why is it important?

It is possible for a component to become disconnected from the Elastic Agent control protocol and unable to complete re-connection because a concurrent update of the component model was occurring.

This change also allows for the runtime manager to not need to settle before calling the initial `m.Update`. You can see that I was able to remove the initial delay in the existing runtime manager tests.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- ~~[ ] I have added an integration test or an E2E test~~

## How to test this PR locally

Only way to really verify is to remove the changes in this PR, except for the `TestManager_FakeInput_NoDeadlock` test. Run the `TestManager_FakeInput_NoDeadlock` test and see that it fails. Then re-add the changes and see that it causes the test to pass.

## Related issues

- Closes #2691

### CODE REVIEW

**Comment by cmacknz** [comment]
No major issues going through this the first time, although I am bit concerned about some of the nested locking we are doing. It is hard for me to reason about whether this will remain deadlock free. The sections where we use the new `currentMx` seem short lived but it is hard for me to see by inspection if there are still possible deadlocks lurking in here.

I did confirm that the new test fails immediately on main but passes here 👍 

Going to do another pass looking at the lock structure before approving.

File: pkg/component/runtime/manager.go [hunk 1]
```diff
@@ -95,7 +95,8 @@ type Manager struct {
 	waitMx    sync.RWMutex
 	waitReady map[string]waitForReady
 
-	mx           sync.RWMutex
+	updateMx     sync.Mutex
```
Comment (fearful-symmetry) [inline pkg/component/runtime/manager.go:98]: Small nit: There's multiple mutexes in this struct now. Can we have some brief comments explaining what they're used or?

Comment (cmacknz) [reply pkg/component/runtime/manager.go:98]: +1, reading the uses of them it became more obvious but adding some comments to avoid misuse is a good idea.

Comment (blakerouse) [reply pkg/component/runtime/manager.go]: Added the requested comments to the mutexes.

**Comment by blakerouse** [comment]
> No major issues going through this the first time, although I am bit concerned about some of the nested locking we are doing. It is hard for me to reason about whether this will remain deadlock free. The sections where we use the new `currentMx` seem short lived but it is hard for me to see by inspection if there are still possible deadlocks lurking in here.

If anything this change reduced the chance, because the use of `currentMx` is always short lived and only for 1 specific purpose now. The old mutex `mx` was used for multiple purposes and was held through the entire update cycle. The the addition of `updateMx` to protect just multiple calls to `update` it allows less lock contention.

**Comment by blakerouse** [comment]
@cmacknz Interesting, how did you run that? I ran `RACE_DETECTOR=true mage unitTest` before proposing this PR. Maybe that didn't work as I was expecting?

**Comment by cmacknz** [comment]
`go test ./pkg/component/... -run TestManager_FakeInput_NoDeadlock -v -count 1 -race`

The command as well as a few of the other race detector findings are in the file I attached.

**Comment by blakerouse** [comment]
@cmacknz Then I would say that `RACE_DETECTOR=true` is not working as it should, and that should be on by default for the CI and `mage unitTest`.

**Comment by blakerouse** [comment]
@cmacknz Fixed the issues detected by the race checker.

**Review by cmacknz — APPROVED** [review]
Thanks, looks good after the latest change.

**Pull request merged**